File: ocaml/lib/ppc/specification.ml
module TemplateMatch = struct
  type rule = Rule

  type t = {
    match_template: string;
    rule: rule option;
    rewrite_template: string option;
  }

  let create rewrite_template rule match_template =
    { match_template; rule; rewrite_template }

  type extracted =
    | Regex of string
    | ContiguousWhitespace of string
    | NonSpace of string

  let escape s =
    let escape_char c =
      if String.contains "\\.|+*?()[]{}^$" c then
        "\\" ^ (String.make 1 c)
      else
        String.make 1 c
    in
    String.concat "" (List.map escape_char (String.to_seq s |> List.of_seq))

  let to_regex t =
    let extracted = extract t.match_template in
    let extracted_to_regex = function
      | Regex s -> s
      | NonSpace s -> escape s
      | ContiguousWhitespace _ -> "\\s+"
    in
    "(" ^ (String.concat "" (List.map extracted_to_regex extracted)) ^ ")"
end

File: rust/src/ppc/specification.rs
pub mod template_match {
    use regex::escape;

    pub struct Rule;

    pub struct T {
        pub match_template: String,
        pub rule: Option<Rule>,
        pub rewrite_template: Option<String>,
    }

    impl T {
        pub fn create(rewrite_template: Option<String>, rule: Option<Rule>, match_template: String) -> Self {
            T {
                match_template,
                rule,
                rewrite_template,
            }
        }
    }

    enum Extracted {
        Regex(String),
        ContiguousWhitespace(String),
        NonSpace(String),
    }

    impl Extracted {
        fn to_regex(&self) -> String {
            match self {
                Extracted::Regex(s) => s.clone(),
                Extracted::ContiguousWhitespace(_) => "\\s+".to_string(),
                Extracted::NonSpace(s) => escape(s),
            }
        }
    }

    pub fn to_regex(t: &T) -> String {
        let extracted = extract(&t.match_template);
        let regex_parts: Vec<String> = extracted.iter().map(|e| e.to_regex()).collect();
        format!("({})", regex_parts.join(""))
    }

    fn extract(template: &str) -> Vec<Extracted> {
        vec![]
    }
}